File: CS502024/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int getDigitEODSL(long int cardN, long int mult);
int getDigitEODL(long int cardN, long int mult);
int getDigitEODSLRaw(long int cardN, long int mult);

int main()
{
    long int cardNumber = get_long("Number: ");
    int cksum1 = 0;
    int cksum2 = 0;
    int cardT = 0; // 1 - Visa | 2 - MasterCard | 3 - AMEX
    long int multiplier = 1;
    for (int i = 0; i < 8; i++)
    { // Picks numbers skiping a decimal place
        cksum1 += getDigitEODSL(cardNumber, multiplier);
        multiplier *= 100;
    }
    multiplier = 1;
    for (int i = 0; i < 8; i++)
    { // Picks numbers skiping a decimal place starting with the last
        cksum2 += getDigitEODL(cardNumber, multiplier);
        multiplier *= 100;
    }
    multiplier /= 100;
    if (getDigitEODSLRaw(cardNumber, multiplier) == 4 ||
        getDigitEODL(cardNumber, multiplier / 100) == 4)
    {
        cardT = 1;
    }
    else if (getDigitEODSLRaw(cardNumber, multiplier) == 5 &&
             getDigitEODL(cardNumber, multiplier) > 0 && getDigitEODL(cardNumber, multiplier) < 6)
    {
        cardT = 2;
    }
    else if (getDigitEODSLRaw(cardNumber, multiplier) == 0 &&
             getDigitEODL(cardNumber, multiplier) == 3 &&
             (getDigitEODSLRaw(cardNumber, multiplier / 100) == 4 ||
              getDigitEODSLRaw(cardNumber, multiplier / 100) == 7))
    {
        cardT = 3;
    }
    if ((cksum1 + cksum2) % 10 == 0)
    {
        if (cardT == 1)
        {
            printf("VISA\n");
        }
        else if (cardT == 2)
        {
            printf("MASTERCARD\n");
        }
        else if (cardT == 3)
        {
            printf("AMEX\n");
        }
        else
            printf("INVALID\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

int getDigitEODSL(long int cardN, long int mult)
{ // Every Other Digit Starting from Second to Last
    int temp = cardN % (100 * mult) / (10 * mult) * 2;
    if (temp >= 10)
    {
        temp = temp % 10 + 1;
    }
    return temp;
}

int getDigitEODL(long int cardN, long int mult)
{ // Every Other Digit Starting from Last
    return cardN % (10 * mult) / (1 * mult);
}

int getDigitEODSLRaw(long int cardN, long int mult)
{ // Every Other Digit Starting from Second to Last
    return cardN % (100 * mult) / (10 * mult);
}

File: CS502024/sentimental-credit/credit.py
import cs50


def main():
    card_num = cs50.get_int("Number: ")
    chk_sum1 = 0
    chk_sum2 = 0
    card_t = 0
    mult = 1
    for i in range(0, 8):
        chk_sum1 += getDigitEODSL(card_num, mult)
        mult *= 100
    mult = 1
    for i in range(0, 8):
        chk_sum2 += getDigitEODL(card_num, mult)
        mult *= 100
    mult /= 100
    if getDigitEODSLRaw(card_num, mult) == 4 or getDigitEODL(card_num, mult / 100) == 4:
        card_t = 1
    elif getDigitEODSLRaw(card_num, mult) == 5 and getDigitEODL(card_num, mult) > 0 and getDigitEODL(card_num, mult) < 6:
        card_t = 2
    elif getDigitEODSLRaw(card_num, mult) == 0 and getDigitEODL(card_num, mult) == 3 and (getDigitEODSLRaw(card_num, mult / 100) == 4 or getDigitEODSLRaw(card_num, mult / 100) == 7):
        card_t = 3
    if (chk_sum1 + chk_sum2) % 10 == 0:
        if card_t == 1:
            print("VISA")
        elif card_t == 2:
            print("MASTERCARD")
        elif card_t == 3:
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")


def getDigitEODSL(card, multp):
    temp = int(card % (100 * multp) / (10 * multp)) * 2
    if temp >= 10:
        temp = temp % 10 + 1
    return temp


def getDigitEODSLRaw(card, multp):
    return int(card % (100 * multp) / (10 * multp))


def getDigitEODL(card, multp):
    return int(card % (10 * multp) / multp)


main()
